### ISSUE TITLE
Making pull-based traversal in dobfs a general operator.

### ISSUE THREAD

**Issue by yzhwang** [gunrock/gunrock#39]
New advance should look like this:
if (meets_some_metric)
  push-based-advance //could be either TWC or LB
  pull-based-advance //again, could be either TWC or LB

Need to be very careful when switching/generating frontiers for pull/push based advance.

**Comment by neoblizz** [comment]
@yzhwang this still needs some attention, GraphIT does a neat job in making push-pull based traversals first class citizens of the programming model, I think Gunrock should definitely consider that as well. In the new API, we have PageRank and BFS (recent pull request) using Push-Pull. If you are interested, this is a really core project that needs a look.

**Comment by yzhwang** [comment]
@neoblizz You mean exposing both push-/pull-based advance as operators right? By using Push-Pull, you mean we have a wrapper where we consider the best iteration to switch between push and pull right? I think we still need that as well, but exposing both push-/pull-based advance sounds like a good thing towards better flexibility. I can start thinking about this. I think I'm very interested in operator-level python binding as well. That said, I probably can only start doing real work in late July. But I will definitely start thinking about this.

**Comment by neoblizz** [comment]
Yes, that's exactly what I am referring to, however, switching from one to another doesn't necessarily have to be automatic- as long as the API is exposed to the user, initially we can assume that they know the kind of implementation they are going for and can switch push-pull accordingly. But eventually, that may be something we control.

Operator-level python binding also sounds great, also worth considering a DSL for high level interface; If you haven't checked out GraphIT, it is worth a look as well https://graphit-lang.org/

No worries, I was just pointing out some interesting issues that are worth a look.

**Comment by jowens** [comment]
@yzhwang Just checkin' on this, I know you hoped to find some time to think about Gunrock. Would love to get the band back together.
